src/image.rs

```diff
@@ -617,6 +618,22 @@ pub trait ImageDecoder<'a>: Sized {
 
         Ok(())
     }
+
+    /// Set decoding limits for this decoder. See [`Limits`] for the different kinds of 
+    /// limits that is possible to set.
+    ///
+    /// Note to implementors: make sure you call [`Limits::check_support`] so that 
+    /// decoding fails if any unsupported strict limits are set.
+    /// [`Limits::check_support`] will also check the `max_image_width` and
+    /// `max_image_height` limits.
+    ///
+    /// [`Limits`]: ./io/struct.Limits.html
+    /// [`Limits::check_support`]: ./io/struct.Limits.html#method.check_support
+    fn set_limits(&mut self, limits: crate::io::Limits) -> ImageResult<()> {
+        limits.check_support(self, crate::io::LimitSupport::default())?;
+
+        Ok(())
+    }
```

Comment:
The `Limits` struct could grow to quite a size, it should be passed by reference if this is feasible. Or did I overlook a good reason to pass it by value?

Also, what other kinds of implementations of this method and in particular the check might there be? It seems odd—but most flexible—that the decoder might be the responsible for performing it. Did you consider the alternative

```rust
/// Enable support for the limit types requested, return support matrix.
fn limit_support(&mut self, limits: &Limits) -> ImageResult<LimitSupport> {
    LimitSupport::default()
}
```

Note that I'm _not_ asking you to change it right away, I'm wondering about tradeoffs of either interface.

Reply:
I'd imagine that a lot of decoders would want to hold onto a copy of limits, which would favor either an owned or boxed value.

Reply:
Passing by reference is probably a good idea, I think most decoders are going to construct their own limit struct and only need access to the values. 

The idea was of course that decoders would enable their own limits internal in `set_limits`, however there is no reason  this can't be done in `limit_support`. I did think about the possibility of having a method of type `fn limit_support(&self) -> LimitSupport`, but did not see enough of a reason to add it. This may be a better alternative.

Reply:
Returning a LimitSupport more error-prone to me, and also kind of like leaking an "internal" implementation detail. I'd prefer if only implementors of ImageDecoder needed to think about forward compatibility of limit restrictions (and the associated LimitSupport type) while callers to set_limits just needed to know that they'd get an unsupported error if they tried to set a limit they weren't allowed to.

Reply:
That is a very good point. I would think external users of this library are far more likely to want to use `set_limits` than they are to implement `ImageDecoder`. If `set_limits` returns the `LimitSupport` it becomes very awkward to use

Reply:
Agreed, `set_limits` appears more ergonomic overall. With regards to copying, decoders will want to hold onto only those limits they support and not the full non-exhaustive struct. And the caller may want to reuse the limits. None of this is really consequential when the `Copy` bounds stay, but otherwise I'd rather have:

> fn set_limits(&mut self, limits: &crate::io::Limits) -> ImageResult<()>

Reply:
Do we ever need to reuse the limits after calling `set_limits` though?

Reply:
Not in the current `io::Reader` but I consider that a more general caller that uses the interface might want to.

Reply:
My only outstanding concern is over deciding between:

* `ImageDecoder { fn limit_support(&mut self, limits: &Limits) }`
* `ImageDecoder { fn limit_support(&mut self, limits: Limits) }`

However, we've agreed on have mutation methods on `Limits`. This makes the first alternative more brittle since it allows a use where the decoder clones its instance while the caller reuses the original instance, effectively duplicating the resources and thus subverting the limit check. So, let's keep it precisely how it was proposed here.